io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/DuplicateIDsFilterIterator.java

```diff
@@ -0,0 +1,159 @@
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution. If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Contributors:
+ *     wetransform GmbH <http://www.wetransform.to>
+ */
+
+package eu.esdihumboldt.hale.io.gml.reader.internal.wfs;
+
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+
+import javax.xml.namespace.QName;
+
+import eu.esdihumboldt.hale.common.instance.model.Instance;
+import eu.esdihumboldt.hale.common.instance.model.ext.InstanceIterator;
+import eu.esdihumboldt.hale.common.schema.model.TypeDefinition;
+import eu.esdihumboldt.hale.io.gml.geometry.GMLConstants;
+
+/**
+ * Filter the instances by gml:id, storing only those with unique IDs.
+ */
+public class DuplicateIDsFilterIterator implements InstanceIterator {
+
+	private final HashSet<String> uniqueIDInstances = new HashSet<String>();
+	private final InstanceIterator iterator;
+	private Instance nextInstance;
+	private boolean instranceAlreadyReturned = false;
+
+	/**
+	 * @param instanceIterator InstanceIterator
+	 */
+	public DuplicateIDsFilterIterator(InstanceIterator instanceIterator) {
+		this.iterator = instanceIterator;
+	}
+
+	/**
+	 * 
+	 * @see eu.esdihumboldt.hale.common.instance.model.ResourceIterator#close()
+	 */
+	@Override
+	public void close() {
+		iterator.close();
+		nextInstance = null;
+		instranceAlreadyReturned = true;
+	}
+
+	/**
+	 * @return TypeDefinition
+	 * @see eu.esdihumboldt.hale.common.instance.model.ext.InstanceIterator#typePeek()
+	 */
+	@Override
+	public TypeDefinition typePeek() {
+		hasNext();
+		if (nextInstance == null) {
+			return null;
+		}
+		return nextInstance.getDefinition();
+	}
+
+	/**
+	 * @return boolean
+	 * @see eu.esdihumboldt.hale.common.instance.model.ext.InstanceIterator#supportsTypePeek()
+	 */
+	@Override
+	public boolean supportsTypePeek() {
+		return true;
+	}
+
+	/**
+	 * 
+	 * @see eu.esdihumboldt.hale.common.instance.model.ext.InstanceIterator#skip()
+	 */
+	@Override
+	public void skip() {
+		next();
+	}
+
+	/**
+	 * @return boolean
+	 * @see java.util.Iterator#hasNext()
+	 */
+	@Override
+	public boolean hasNext() {
+		if (nextInstance == null || instranceAlreadyReturned) {
+
+			while (true) {
+				if (iterator != null && !iterator.hasNext()) {
+					return false;
+				}
+				else {
+					Instance instance = iterator.next();
+					if (!isTheInstancePresent(instance)) {
+						nextInstance = instance;
+						instranceAlreadyReturned = false;
+						return true;
+					}
+				}
+			}
+		}
+		else {
+			return true;
+		}
+	}
+
+	/**
+	 * @return Instance
+	 * @see java.util.Iterator#next()
+	 */
+	@Override
+	public Instance next() {
+		hasNext();
+		if (nextInstance == null) {
+			for (String string : uniqueIDInstances) {
+				System.out.println(string);
+			}
+			throw new NoSuchElementException();
+		}
+		instranceAlreadyReturned = true;
+		return nextInstance;
+	}
+
+	private boolean isTheInstancePresent(Instance instance) {
```

Comment:
I think it would help to add documentation to the method to easily understand what it is supposed to do.